File: Stack/150. Evaluate Reverse Polish Notation\/Solution.java
import java.util.*;

class Solution {
    private int evalute(int num2, int num1, String operand){
        if (operand.equals("+")) return num1 + num2;
        else if (operand.equals("-")) return num1 - num2;
        else if (operand.equals("*")) return num1 * num2;
        else return num1 / num2;
    }

    private int dfs(Stack<String> tokens, HashSet<String> operands){
        String token = tokens.pop();
        if (!operands.contains(token)) return Integer.parseInt(token);

        int num1 = dfs(tokens, operands);
        int num2 = dfs(tokens, operands);
        return evalute(num1, num2, token);
    }
    public int evalRPN(String[] tokens) {
        // N: len(tokens)
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        // Stack<String> tokenList = new Stack<>();
        // HashSet<String> operands = new HashSet<>(List.of("+", "-", "*", "/"));
        // tokenList.addAll(Arrays.asList(tokens));
        // return dfs(tokenList, operands);


        // N: len(tokens)
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        Stack<Integer> evalStack = new Stack<>();
        HashSet<String> operands = new HashSet<>(List.of("+", "-", "*", "/"));
        
        for (String token: tokens){
            if (!operands.contains(token)) evalStack.push(Integer.parseInt(token));
            else evalStack.push(evalute(evalStack.pop(), evalStack.pop(), token));
        }
        return evalStack.peek();
    }
}

File: Stack/150. Evaluate Reverse Polish Notation\/Solution.py
from typing import List

class Solution:
    def evaluate(self, num2: int, num1: int, operand: str) -> int:
        if operand == '+':
            return num1 + num2
        elif operand == '-':
            return num1 - num2
        elif operand == '*':
            return num1 * num2
        elif operand == '/':
            return int(num1 / num2)
        
    def evalRPN(self, tokens: List[str]) -> int:
        # N: len(tokens)
        # Time Complexity: O(N)
        # Space Complexity: O(N) -> Stack Space
        # s = Solution()
        # operand = '+-*/'

        # def dfs():
        #     token = tokens.pop()
        #     if token not in operand:
        #         return int(token)
            
        #     num1 = dfs()
        #     num2 = dfs()
        #     return s.evaluate(num1, num2, token)
        # return dfs()


        # N: len(tokens)
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        s = Solution()
        operand = '+-*/'
        stack = []

        for token in tokens:
            if token not in operand:
                stack.append(int(token))
            else:
                stack.append(s.evaluate(stack.pop(), stack.pop(), token))
        return stack[-1]